File: java/src/main/java/com/workshop/architecture/fitness/create_customer/CreateCustomerController.java
package com.workshop.architecture.fitness.create_customer;

import com.workshop.architecture.fitness.shared.CustomerEntity;
import com.workshop.architecture.fitness.shared.CustomerRepository;
import com.workshop.architecture.fitness.shared.CustomerResponse;
import com.workshop.architecture.fitness.shared.CustomerUpsertRequest;
import jakarta.validation.Valid;
import java.net.URI;
import java.util.UUID;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/api/customers")
public class CreateCustomerController {

    private final CustomerRepository customerRepository;

    public CreateCustomerController(CustomerRepository customerRepository) {
        this.customerRepository = customerRepository;
    }

    @PostMapping
    ResponseEntity<CustomerResponse> createCustomer(@Valid @RequestBody CustomerUpsertRequest request) {
        CustomerEntity customer = new CustomerEntity(
                UUID.randomUUID(),
                request.name(),
                request.dateOfBirth(),
                request.emailAddress()
        );

        CustomerResponse response = CustomerResponse.fromEntity(customerRepository.save(customer));
        return ResponseEntity.created(URI.create("/api/customers/" + response.id())).body(response);
    }
}

File: python/src/workshop_api/fitness/managing_customers/create_customer/create_customer_controller.py
from uuid import uuid4

from fastapi import APIRouter, Depends, Response, status
from sqlalchemy.orm import Session

from workshop_api.fitness.database import get_db_session
from workshop_api.fitness.managing_customers.shared.customer_entity import CustomerOrmModel
from workshop_api.fitness.managing_customers.shared.customer_response import CustomerResponse
from workshop_api.fitness.managing_customers.shared.customer_upsert_request import (
    CustomerUpsertRequest,
)

router = APIRouter(prefix="/api/customers", tags=["customer"])


@router.post(
    "",
    response_model=CustomerResponse,
    response_model_by_alias=True,
    status_code=status.HTTP_201_CREATED,
)
def create_customer(
    request: CustomerUpsertRequest,
    response: Response,
    session: Session = Depends(get_db_session),
) -> CustomerResponse:
    customer = CustomerOrmModel(
        id=str(uuid4()),
        name=request.name,
        date_of_birth=request.date_of_birth,
        email_address=request.email_address,
    )

    session.add(customer)
    session.commit()
    session.refresh(customer)

    created = CustomerResponse.from_entity(customer)
    response.headers["Location"] = f"/api/customers/{created.id}"
    return created
